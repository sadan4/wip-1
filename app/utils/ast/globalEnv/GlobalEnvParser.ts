import { isArrayLiteralExpression, isBigIntLiteral, isBinaryExpression, isNumericLiteral, isObjectLiteralExpression, isPropertyAssignment, isStringLiteral, isStringLiteralLike, SyntaxKind, type Expression, type LiteralToken, type ObjectLiteralElement, type ObjectLiteralExpression } from "typescript";
import { AstParser } from "../AstParser";
import { Logger } from "../../Logger";
import { Cache, CacheGetter } from "../../decorators"

interface NonLiteral {
    [GlobalEnvParser.SYM_UNSERIALIZABLE]: true;
    expression: string;
}

const logger = new Logger("GlobalEnvParser");

export type MaybeLiteralJsonType = JsonType | NonLiteral;
export type JsonType = string | number | boolean | null | NonLiteral | MaybeLiteralJsonType[] | { [key: string]: MaybeLiteralJsonType }
export type EnvBuildVars = Record<string, JsonType> & { [GlobalEnvParser.SYM_UNREADABLE_KEYS]: string[] }

export class GlobalEnvParser extends AstParser {
    static readonly SYM_UNSERIALIZABLE = Symbol("GlobalEnvParser.unserializable");
    static readonly SYM_UNREADABLE_KEYS = Symbol("GlobalEnvParser.unreadableKeys");

    public static isLiteral(value: MaybeLiteralJsonType): value is Exclude<JsonType, NonLiteral> {
        if (value === null || typeof value !== "object") {
            return true;
        }
        return !(GlobalEnvParser.SYM_UNSERIALIZABLE in value);
    }

    public static isObject(value: MaybeLiteralJsonType): value is { [key: string]: MaybeLiteralJsonType } {
        if (!GlobalEnvParser.isLiteral(value)) {
            return false;
        }
        return typeof value === "object" && value !== null && !Array.isArray(value);
    }

    public static isSingleValue(value: MaybeLiteralJsonType): value is string | number | boolean | null {
        if (!GlobalEnvParser.isLiteral(value)) {
            return false;
        }
        if (Array.isArray(value)) {
            return false;
        }
        if (typeof value === "object" && value !== null) {
            return false;
        }
        return true;
    }

    @CacheGetter()
    public get buildId(): string | undefined {
        const { SENTRY_TAGS } = this.getGlobalEnvObject() ?? {} as EnvBuildVars;
        if (typeof SENTRY_TAGS !== "object" || !GlobalEnvParser.isObject(SENTRY_TAGS)) {
            return;
        }

        const maybeBuildId = SENTRY_TAGS?.buildId;

        if (typeof maybeBuildId !== "string") {
            return;
        }
        return maybeBuildId;
    }

    @Cache()
    public getGlobalEnvObject(): EnvBuildVars | undefined {
        if (!this.text.includes("window.GLOBAL_ENV")) {
            logger.error("expected text to contain 'window.GLOBAL_ENV'");
            return undefined;
        }
        const expr = this.sourceFile.getChildAt(0)?.getChildAt(0)?.getChildAt(0);
        if (!isBinaryExpression(expr)) {
            logger.error("expected global env to contain a top-level binary expression");
            return;
        }

        if (expr.operatorToken.kind !== SyntaxKind.EqualsToken) {
            logger.error(`expected assignment operator, got ${expr.operatorToken.getText()}`);
            return;
        }

        if (!isObjectLiteralExpression(expr.right)) {
            logger.error("expected an object literal to be assigned to window.GLOBAL_ENV");
            return;
        }

        return this.parseObjectLiteral(expr.right);
    }

    private parseLiteral(node: LiteralToken) {
        if (isStringLiteralLike(node) || isBigIntLiteral(node)) {
            return node.text;
        } else if (isNumericLiteral(node)) {
            return Number(node.text);
        } else {
            throw new Error("weird unsupported literal token in globalEnv");
        }
    }

    private parseValue(node: Expression): [JsonType, string[]] {
            if (this.isLiteralish(node)) {
                return [this.parseLiteral(node), []];
            } else if (isObjectLiteralExpression(node)) {
                const parsed: Omit<EnvBuildVars, typeof GlobalEnvParser.SYM_UNREADABLE_KEYS> = this.parseObjectLiteral(node);
                const keys = [...(parsed as any)[GlobalEnvParser.SYM_UNREADABLE_KEYS]]

                delete (parsed as any)[GlobalEnvParser.SYM_UNREADABLE_KEYS];

                return [parsed, keys];
            } else if (isArrayLiteralExpression(node)) {
                const parsed = node.elements.map(this.parseValue.bind(this)).reduce((acc, [jsonvalue, badkeys]) => {
                    // @ts-expect-error loses the [] from JsonType[] for some reason
                    acc[0].push(jsonvalue)
                    acc[1].push(...badkeys)
                    return acc;
                }, [[], []] as [JsonType[], string[]]);
                return parsed;
            } else {
                return [{[GlobalEnvParser.SYM_UNSERIALIZABLE]: true, expression: node.getText()}, []];
            }
    }

    private parseObjectLiteral(expr: ObjectLiteralExpression): EnvBuildVars {
        const ret: EnvBuildVars = { [GlobalEnvParser.SYM_UNREADABLE_KEYS]: [] };
        for (const prop of expr.properties) {
            if (!prop.name || !isStringLiteralLike(prop.name) || !isPropertyAssignment(prop)) {
                ret[GlobalEnvParser.SYM_UNREADABLE_KEYS].push(prop.name?.getText() ?? "<unknown>");
                continue;
            }
            const [value, badkeys] = this.parseValue(prop.initializer);
            ret[prop.name.text] = value;
            ret[GlobalEnvParser.SYM_UNREADABLE_KEYS].push(...badkeys);
        }
        return ret;
    }

}
