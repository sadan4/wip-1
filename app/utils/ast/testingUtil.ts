import { readFileSync } from "node:fs";
import { join } from "node:path";

export function makeGetFile(__dirname: string) {
    return function getFile(asset: string): string {
        return readFileSync(join(__dirname, "__test__", asset), "utf-8");
    }
}
