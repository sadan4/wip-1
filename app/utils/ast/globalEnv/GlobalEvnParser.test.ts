import { describe, expect, it } from "vitest";
import { makeGetFile } from "../testingUtil";
import { GlobalEnvParser } from "./GlobalEnvParser";

const getFile = makeGetFile(import.meta.dirname);

describe("GlobalEnvParser", () => {
    it("parses an example discord buildenv correctly", () => {
        const parser = new GlobalEnvParser(getFile("exampleEnv.js"));
        const env = parser.getGlobalEnvObject();
        expect(env).toMatchSnapshot();
    })
});
