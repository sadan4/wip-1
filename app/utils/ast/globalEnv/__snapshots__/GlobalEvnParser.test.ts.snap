// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GlobalEnvParser > parses an example discord buildenv correctly 1`] = `
{
  "ACTIVITY_APPLICATION_HOST": "discordsays.com",
  "ADYEN_KEY": "live_E3OQ33V6GVGTXOVQZEAFQJ6DJIDVG6SY",
  "API_ENDPOINT": "//canary.discord.com/api",
  "API_PROTOCOL": "https:",
  "API_VERSION": 9,
  "ASSET_ENDPOINT": "//canary.discord.com",
  "BRAINTREE_KEY": "production_ktzp8hfp_49pp2rp4phym7387",
  "BUILD_NUMBER": "440758",
  "BUILT_AT": "1757033646091",
  "CDN_HOST": "cdn.discordapp.com",
  "DEVELOPERS_ENDPOINT": "//canary.discord.com",
  "DEV_SESSION_KEY": "undefined",
  "GATEWAY_ENDPOINT": "wss://gateway.discord.gg",
  "GIFT_CODE_HOST": "discord.gift",
  "GUILD_TEMPLATE_HOST": "discord.new",
  "HTML_TIMESTAMP": {
    "expression": "Date.now()",
    Symbol(GlobalEnvParser.unserializable): true,
  },
  "IMAGE_PROXY_ENDPOINTS": "//images-ext-1.discordapp.net,//images-ext-2.discordapp.net",
  "INVITE_HOST": "discord.gg",
  "LOCATION": "history",
  "MARKETING_ENDPOINT": "//canary.discord.com",
  "MEDIA_PROXY_ENDPOINT": "//media.discordapp.net",
  "MIGRATION_DESTINATION_ORIGIN": "https://canary.discord.com",
  "MIGRATION_SOURCE_ORIGIN": "https://canary.discordapp.com",
  "NETWORKING_ENDPOINT": "//router.discordapp.net",
  "NODE_ENV": "production",
  "PRIMARY_DOMAIN": "discord.com",
  "PROJECT_ENV": "production",
  "PUBLIC_PATH": "/assets/",
  "RELEASE_CHANNEL": "canary",
  "REMOTE_AUTH_ENDPOINT": "//remote-auth-gateway.discord.gg",
  "RTC_LATENCY_ENDPOINT": "//latency.discord.media/rtc",
  "SENTRY_RELEASE": "discord_web-a91a4c21106f49567ccca288e233457b019fd4f4",
  "SENTRY_TAGS": {
    "buildId": "a91a4c21106f49567ccca288e233457b019fd4f4",
    "buildType": "normal",
  },
  "SEO_ENDPOINT": "undefined",
  "SPRIG_API_KEY": "ZaQ2JuStvc",
  "STATIC_ENDPOINT": "",
  "STRIPE_KEY": "pk_live_CUQtlpQUF0vufWpnpUmQvcdi",
  "VERSION_HASH": "a91a4c21106f49567ccca288e233457b019fd4f4",
  "WEBAPP_ENDPOINT": "//canary.discord.com",
  "WIDGET_ENDPOINT": "//canary.discord.com/widget",
  Symbol(GlobalEnvParser.unreadableKeys): [],
}
`;
